feat: middleware for several adapters (#13477)
sveltejs/kit: documentation/docs/25-build-and-deploy/90-adapter-vercel.md

Vercel supports [Incremental Static Regeneration](https://vercel.com/docs/incremental-static-regeneration) (ISR), which provides the performance and cost advantages of prerendered content with the flexibility of dynamically rendered content.

> > [!NOTE] Use ISR only on routes where every visitor should see the same content (much like when you prerender). If there's anything user-specific happening (like session cookies), they should happen on the client via JavaScript only to not leak sensitive information across visits

Reviewer: ```suggestion
> [!WARNING] Use ISR only on routes where every visitor should see the same content (much like when you prerender). If there's anything user-specific happening (like session cookies), they should happen on the client via JavaScript only to not leak sensitive information across visits
```

seems wise

A list of valid query parameters that contribute to the cache key. Other parameters (such as utm tracking codes) will be ignored, ensuring that they do not result in content being re-generated unnecessarily. By default, query parameters are ignored.

> [!NOTE] Pages that are  [prerendered](page-options#prerender) will ignore ISR configuration.

## Edge Middleware

> You can make use of [Vercel Edge Middleware](https://vercel.com/docs/functions/edge-middleware) by placing an `edge-middleware.js` file at the root of your project. You can use it to intercept requests even for prerendered or ISR'd pages. Combined with using [server-side route resolution](configuration#router) you can make sure it runs prior to all navigations, no matter client- or server-side. This allows you to for example run A/B-tests on prerendered or ISR'd pages by rerouting a user to either variant A or B depending on a cookie.

Reviewer: ```suggestion
You can make use of [Vercel Edge Middleware](https://vercel.com/docs/functions/edge-middleware) by placing an `edge-middleware.js` file at the root of your project. You can use it to intercept all requests including those for ISRed pages and static content. Combined with [server-side route resolution](configuration#router), you can ensure it runs prior to all navigations, whether client- or server-side. This allows you to, for example, run A/B tests on prerendered or ISRed pages by rerouting a user to either variant A or B depending on a cookie.
```

Stylistic / grammar

```js
/// file: edge-middleware.js
// @filename: ambient.d.ts
declare module '@vercel/edge';

// @filename: index.js
// ---cut---

> import { rewrite, next } from '@vercel/edge';

Reviewer: Do rewrites actually work in SvelteKit yet?

Author: Yes they do, because we wrap the middleware behind the scenes to make it work

/**
 * @param {Request} request
 */
export default async function middleware(request) {
	const url = new URL(request.url);

	if (url.pathname !== '/') return next();

	// Retrieve feature flag from cookies
	let flag = split_cookies(request.headers.get('cookie') ?? '')?.flag;

	// Fall back to random value if this is a new visitor
	flag ||= Math.random() > 0.5 ? 'a' : 'b';

	return rewrite(
		// Get destination URL based on the feature flag
		flag === 'a' ? '/home-a' : '/home-b',
		{
			headers: {
				// Set a cookie to remember the feature flags for this visitor
				'Set-Cookie': `flag=${flag}; Path=/`
			}
		}
	);
}

/** @param {string} cookies */
function split_cookies(cookies) {
	return cookies.split(';').reduce(
		(acc, cookie) => {
			const [name, value] = cookie.trim().split('=');
			acc[name] = value;
			return acc;
		},
		{} as Record<string, string>
	);
}
```

> By default, middleware runs on all requests except for files within `_app/immutable`. You can customize this by exporting a `config` object with a `matcher` property as described in Vercel's [API documentation](https://vercel.com/docs/functions/edge-middleware/middleware-api#match-paths-based-on-custom-matcher-config).

Reviewer: Random comment, but... do people actually know what stuff is in `_app/immutable`? I don't know if it's an easy explanation but I feel like it's kind of arcane knowledge

> > [!NOTE] If you want to run code prior to a request but neither have prerendered nor ISR'd pages and have no rerouting logic, then it makes more sense to use the [handle hook](hooks#Server-hooks-handle) instead.

Reviewer: ```suggestion
> [!NOTE] If you can do what you need to by using the [handle hook](hooks#Server-hooks-handle), do so. Edge middleware has a not-insignificant performance penalty and should only be used if absolutely necessary. Notable use cases for Edge Middleware include A/B testing using static content and rewrites, or running lightweight logic (such as adding headers) while serving ISRed content.
```

Same as Netlify, put the warning first
